### ISSUE TITLE
Hash documentation needs to be updated to reflect all methods of creating Hash's and their rules.

### ISSUE THREAD

**Issue by samcv** [perl6/doc#1380]
See this [StackOverflow question](https://stackoverflow.com/questions/44101411/is-that-a-perl-6-hash-or-block/44104556?noredirect=1#comment75267000_44104556) and my answer at the bottom for background.

Using `{ }` to create a Hash in the Perl 5 way. While the Synopsis makes it clear that `%()` is strongly preferred for creating Hash's, and `{ }` is a conveinence backcompatibility, the documentation is not as strongly worded.

The [Hash](https://docs.perl6.org/type/Hash) docs page mainly uses the `{ }` and then tells you to use many tricks to try and avoid the pitfalls of using these, when it sometimes will create a Block not a Hash.

Firstly the Hash page needs to be updated to use `%( )` extensively and saying that `{ }` is provided for convenience. We can explain the pitfalls as a **reason** to always use `%( )`, instead of explaining what to do *if* `{ }` doesn't work.

Secondly, we need to update the rest of the docs to also use `%( )`, in places where it makes sense.

**Comment by JJ** [comment]
Maybe add a test for this?

**Comment by AlexDaniel** [comment]
@JJ I think just a mention in CONTRIBUTING.md would be enough.

**Comment by nxadm** [comment]
>  We can explain the pitfalls as a reason to always use %( ), instead of explaining what to do if { } doesn't work.

If the reasoning is to always recommend %( ) (I agree), would it be better to obsolete { } for hashes? It could warn for some releases to be removed in a few years. Otherwise people will have to know about { } hashes *and* their pitfalls when reading code, even if they don't use it.

**Comment by AlexDaniel** [comment]
@nxadm there was some discussion on that topic in https://github.com/rakudo/rakudo/issues/1644

TL;DR: the empty block/hash pitfall can be detrapped by giving a warning for the empty case (`{}`), but generally curly braces for hashes will have to stay.

**Comment by raiph** [comment]
Hi @samcv, I'm pondering changes to doc related to `%()` and `{}`.

(Hopefully no one other than you is pinged by this comment because the thread is closed and we don't need to discuss anything much yet -- I'm just hoping for a single comment by you to help me move forward.)

Per [%(), @(), $() are evil/weird](https://github.com/rakudo/rakudo/issues/1946), I feel that, at the very minimum, someone needs to edit how the docs talk about the `%()` hash constructor. I'm currently planning to do that.

I want to find a way to honor everyones' views about `$( )` vs `{}`. It's going to be difficult to do that but I'm up for the challenge. We can always revert any edits I do end up making if they don't make the grade, but I want to try refresh my understanding of various peoples's view on this topic before I even start making edits.

To develop a better sense of your view, I've just been revisiting some of your comments. I was deeply surprised by this one:

> the Synopsis makes it clear that `%()` is strongly preferred for creating Hash's, and `{ }` is a conveinence backcompatibility

I had not gotten that impression from the synopses. My own view was pretty much in line with what jnthn wrote in the evil thread I linked above that included:

> I don't think the %(...) coercion form was originally intended to become a widely used alternative to {...}

Anyhoo, I began to do a quick search of s02, s03, some apos and exegeses for `%(`. That didn't yield anything one way or the other. And then githubusercontent went offline and still is as I write this. So I thought, well, I need to say hi to you at the very least before I think about editing the docs in relation to this thorny topic. So, hi. :) Plus, can you give me some pointers in the spec that illustrate what you meant by `%()` being strongly preferred (presuming they're working for you)? TIA.

**Comment by AlexDaniel** [comment]
@raiph Some IRC comments by TimToady: http://colabti.org/irclogger/irclogger_log/perl6?date=2018-07-08#l140

**Comment by JJ** [comment]
I don't think the issue is solved, if only because all the comments after it was (theoretically) solved.
